proposals/0370-modifiers.rst

```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
+  ``Multiplicity`` on the same arrow are not allowed. Any other use of a
+  modifier still has no meaning.
+
+* The ``%1`` modifier is handled as a special case. It's renamed (and so
+  typechecked) the same as ``%One``, even if it appears somewhere that linear
+  modifiers aren't expected. If a user does want the modifier ``1 :: Nat``, they
+  can write it as ``%01``. (Requires ``-XDataKinds``.)
+
+* The linear arrow ``a ⊸ b`` has the same meaning as ```a %1 -> b``. Other
+  modifiers are accepted: ``a %Matchable ⊸ b`` has the same meaning as
+  ``a %Matchable %1 -> b``.
+
+With ``-XNoLinearTypes``, the ``%1`` modifier is not special. It refers to the
+type ``1 :: Nat`` and requires ``-XDataKinds``.
+
+``-XLinearTypes`` implies ``-XModifiers``. But the latter can be explicitly
+disabled with ``-XLinearTypes -XNoModifiers``. This introduces backwards
+compatible behavior:
+
+* Only ``Multiplicity`` modifiers are permitted, and only in the places they're
+  recognized. Any use of a modifier is an error.
+
+* The kind of a modifier is determined by checking for ``Multiplicity``, not
+  through synthesis. So ``Int %m -> Bool`` is forbidden with ``-XLinearTypes
+  -XModifiers``, because ``m`` has unknown kind. But it's permitted with
+  ``-XLinearTypes -XNoModifiers``, equivalently to ``Int %(m :: Multiplicity) ->
+  Bool``.
+
+This behavior may be deprecated in future.
+
+Renaming and typechecking
+-------------------------
+Initially, the only recognized modifiers will be the multiplicity modfifiers of
+linear types. These take effect during typechecking. Other modifiers are
+expected to take effect during renaming, such as the ``NoFieldSelectors``
+modifier of `proposal 512`_.
+
+.. _`proposal 512`: https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0512-nofieldselectors-per-datatype.md
+
+This poses a problem. The following is accepted::
+
+  type family F a where
+    F 1 = Many
+    F 2 = NoFieldSelectors
+
+  f :: () %(F 1) -> ()
+
+Ideally, the following would be accepted as well::
+
+  %(F 2) data G = G { g :: () }
+
+But type family resolution happens in the typechecker.
+
+Define an **RN modifier** as one that takes effect during renaming. This is
+context-dependent: ``%NoFieldSelectors`` will be an RN modifier before a
+``data`` declaration, but not before an arrow.
+
+One possible solution is to simply forbid such things for RN modifiers. If the
+type can't be resolved during renaming, it has no effect during renaming. During
+typechecking, if it gets resolved to something that would have had effect during
+renaming, we throw an error.
+
+This isn't ideal because it's not obvious to users which modifiers are RN
+modifiers.
+
+A more complicated solution involves invoking the typechecker during renaming.
+To do this we need a "stage restriction": a modifier obeys the stage restriction
+if every identifier and every type family instance in the modifier is imported,
+not defined in the current module.
```

Comment:
I'm not sure if this stage restriction is standard when dealing with TH or specific to modifiers. IIRC you can do `$(pure ())` at the top level to break a module into two compilation units. Would we expect that to work here, too?

Can we tell whether a type family instance comes from this module or not? Or would it be "we try to typecheck, and if it needs a type family instance that's not defined outside this module, that fails"?